File: Front/lib/models/avis.dart
class Avis {
  final String id;
  final String touristeId;
  final String? touristeFullname;
  final String? touristeAvatar;
  final String? activiteId;
  final String? organisateurId;
  final String type; // 'activite' | 'organisateur'
  final double note;
  final String? commentaire;
  final DateTime createdAt;

  Avis({
    required this.id,
    required this.touristeId,
    this.touristeFullname,
    this.touristeAvatar,
    this.activiteId,
    this.organisateurId,
    required this.type,
    required this.note,
    this.commentaire,
    required this.createdAt,
  });

  factory Avis.fromJson(Map<String, dynamic> json) {
    String touristeId = '';
    String? touristeFullname;
    String? touristeAvatar;

    if (json['touriste_id'] is String) {
      touristeId = json['touriste_id'];
    } else if (json['touriste_id'] is Map) {
      touristeId = json['touriste_id']['_id'] ?? '';
      touristeFullname = json['touriste_id']['fullname'];
      touristeAvatar = json['touriste_id']['avatar'];
    }

    String? activiteId;
    if (json['activite_id'] is String) {
      activiteId = json['activite_id'];
    } else if (json['activite_id'] is Map) {
      activiteId = json['activite_id']['_id'];
    }

    String? organisateurId;
    if (json['organisateur_id'] is String) {
      organisateurId = json['organisateur_id'];
    } else if (json['organisateur_id'] is Map) {
      organisateurId = json['organisateur_id']['_id'];
    }

    return Avis(
      id: json['_id'] ?? '',
      touristeId: touristeId,
      touristeFullname: touristeFullname,
      touristeAvatar: touristeAvatar,
      activiteId: activiteId,
      organisateurId: organisateurId,
      type: json['type'] ?? 'activite',
      note: (json['note'] ?? 0).toDouble(),
      commentaire: json['commentaire'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
    );
  }
}

File: Back/models/avis.js
const mongoose = require("mongoose");

const avisSchema = new mongoose.Schema(
  {
    // Tourist who submitted this review/rating
    touriste_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Touriste",
      required: true,
    },
    // Activity reference (for activity reviews)
    activite_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Activite",
      default: null,
    },
    // Organizer reference (for organizer ratings)
    organisateur_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Organisator",
      default: null,
    },
    // Type of review: activity review or organizer rating
    type: {
      type: String,
      enum: ["activite", "organisateur"],
      required: true,
    },
    // Star rating 1-5
    note: {
      type: Number,
      required: true,
      min: [1, "Note minimum is 1"],
      max: [5, "Note maximum is 5"],
    },
    // Optional text comment
    commentaire: {
      type: String,
      maxlength: [1000, "Comment cannot exceed 1000 characters"],
      default: null,
    },
    // Optional tags (max 3)
    tags: {
      type: [String],
      default: [],
      validate: {
        validator: function(tags) {
          return tags.length <= 3;
        },
        message: "Maximum 3 tags allowed",
      },
    },
    // Link to the inscription that qualifies the tourist to review
    inscription_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Inscription",
    },
  },
  {
    timestamps: true,
  },
);

// One review per tourist per activity
avisSchema.index(
  { touriste_id: 1, activite_id: 1 },
  {
    unique: true,
    partialFilterExpression: {
      type: "activite",
      activite_id: { $ne: null },
    },
  },
);

// One rating per tourist per organizer
avisSchema.index(
  { touriste_id: 1, organisateur_id: 1 },
  {
    unique: true,
    partialFilterExpression: {
      type: "organisateur",
      organisateur_id: { $ne: null },
    },
  },
);

avisSchema.index({ activite_id: 1, type: 1 });
avisSchema.index({ organisateur_id: 1, type: 1 });

const Avis = mongoose.model("Avis", avisSchema);

module.exports = Avis;
